File: app/src/main/java/com/github/overpass/gather/model/usecase/login/SignInUseCase.java
package com.github.overpass.gather.model.usecase.login;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.Transformations;

import com.github.overpass.gather.model.commons.LiveDataUtils;
import com.github.overpass.gather.model.data.validator.BaseValidator;
import com.github.overpass.gather.model.repo.pref.PreferenceRepo;
import com.github.overpass.gather.screen.auth.login.SignInStatus;
import com.github.overpass.gather.model.repo.login.AuthRepo;
import com.github.overpass.gather.screen.splash.StartStatus;

public class SignInUseCase {

    private final AuthRepo authRepo;
    private final BaseValidator validator;
    private final PreferenceRepo preferenceRepo;

    public SignInUseCase(AuthRepo authRepo,
                         BaseValidator credsValidator,
                         PreferenceRepo preferenceRepo) {
        this.authRepo = authRepo;
        this.validator = credsValidator;
        this.preferenceRepo = preferenceRepo;
    }

    public LiveData<SignInStatus> signIn(String email, String password) {
        if (!validator.isEmailValid(email)) {
            return LiveDataUtils.just(new SignInStatus.InvalidEmail("Invalid Email"));
        }
        if (!validator.isPasswordValid(password)) {
            return LiveDataUtils.just(new SignInStatus.InvalidPassword("Invalid Password"));
        }
        return Transformations.map(authRepo.signIn(email, password), status -> {
            preferenceRepo.setStartStatus(StartStatus.AUTHORIZED);
            return status;
        });
    }
}

File: app/src/main/java/com/github/overpass/gather/model/usecase/login/SignInUseCase.kt
package com.github.overpass.gather.model.usecase.login

import com.github.overpass.gather.di.EMAIL_VALIDATOR
import com.github.overpass.gather.di.PASSWORD_VALIDATOR
import com.github.overpass.gather.model.data.entity.signin.SignInStatus
import com.github.overpass.gather.model.data.entity.splash.StartStatus
import com.github.overpass.gather.model.data.validator.Validator
import com.github.overpass.gather.model.repo.login.AuthRepo
import com.github.overpass.gather.model.repo.pref.PreferenceRepo
import com.google.android.gms.tasks.Task
import com.google.android.gms.tasks.Tasks
import javax.inject.Inject
import javax.inject.Named

class SignInUseCase @Inject constructor(
        private val authRepo: AuthRepo,
        @Named(EMAIL_VALIDATOR) private val emailValidator: Validator<String>,
        @Named(PASSWORD_VALIDATOR) private val passwordValidator: Validator<String>,
        private val preferenceRepo: PreferenceRepo
) {

    fun signIn(email: String, password: String): Task<SignInStatus> {
        if (!emailValidator.isValid(email)) {
            return Tasks.forResult(SignInStatus.InvalidEmail("Invalid Email"))
        }
        if (!passwordValidator.isValid(password)) {
            return Tasks.forResult(SignInStatus.InvalidPassword("Invalid Password"))
        }
        return authRepo.signIn(email, password)
                .onSuccessTask {
                    preferenceRepo.startStatus = StartStatus.AUTHORIZED
                    Tasks.forResult(it!!)
                }
    }
}
